File: src/commands/stats.ts
import type { Command } from "commander";
import { hrAtPace, splitShape, splitTable } from "../analysis.ts";
import { loadConfig } from "../config.ts";
import { formatMeters } from "../display.ts";
import { printJSON } from "../envelope.ts";
import { formatSeconds } from "../models.ts";
import { dataPaths } from "../paths.ts";
import type { GoalProgress } from "../stats.ts";
import {
  buildWeekSummaries,
  computeGoalProgress,
  localYMD,
  recentWeeks,
  weekSummaryData,
} from "../stats.ts";
import { readWorkouts } from "../storage.ts";
import { resolveWorkout } from "./show.ts";

export interface GoalProjection {
  projected_total_meters: number;
  projected_pct: number;
  shortfall_meters: number;
}

export function projectGoal(goal: GoalProgress): GoalProjection {
  const remaining = Math.max(0, goal.totalWeeks - goal.weeksElapsed);
  const projected = Math.round(goal.currentAvgPace * remaining + goal.totalMeters);
  return {
    projected_total_meters: projected,
    projected_pct: Math.round((projected / goal.target) * 1000) / 10,
    shortfall_meters: Math.max(0, goal.target - projected),
  };
}

function parseWeeks(raw: string): number {
  const weeks = parseInt(raw, 10);
  if (Number.isNaN(weeks) || weeks < 1) {
    console.error("Error: --weeks must be a positive integer.");
    process.exit(1);
  }
  return weeks;
}

export function registerStats(program: Command): void {
  const stats = program.command("stats").description("Derived training statistics");

  stats
    .command("weekly")
    .description("Weekly volume, sessions, pace, SPM, and HR")
    .option("-w, --weeks <n>", "number of weeks", "12")
    .option("--json", "output as JSON")
    .action(async (opts: { weeks: string; json?: boolean }) => {
      const weeks = parseWeeks(opts.weeks);
      const cfg = await loadConfig();
      const workouts = await readWorkouts(dataPaths(cfg));
      const summaries = buildWeekSummaries(workouts, new Date(), weeks).map(weekSummaryData);

      if (opts.json) {
        printJSON("c2.stats.weekly.v1", { weeks: summaries });
        return;
      }
      console.log("week        meters  sess  pace/500m   spm    hr");
      for (const s of summaries) {
        console.log(
          `${s.week_start}  ${formatMeters(s.meters).padStart(8)}  ${String(s.sessions).padStart(4)}  ${(s.avg_pace_500m ?? "-").padStart(9)}  ${String(s.avg_spm ?? "-").padStart(4)}  ${String(s.avg_hr ?? "-").padStart(4)}`,
        );
      }
    });

  stats
    .command("goal")
    .description("Goal trajectory and projection")
    .option("--json", "output as JSON")
    .action(async (opts: { json?: boolean }) => {
      const cfg = await loadConfig();
      if (!cfg.goal.start_date || !cfg.goal.end_date) {
        console.error("Goal dates not configured. Run `c2 setup` to set start and end dates.");
        process.exit(1);
      }
      const workouts = await readWorkouts(dataPaths(cfg));
      const now = new Date();
      const goal = computeGoalProgress(workouts, cfg, now);
      const projection = projectGoal(goal);
      const weeks = recentWeeks(workouts, now, 4);
      const thisWeek = weeks[0]!;

      if (opts.json) {
        printJSON("c2.stats.goal.v1", {
          goal,
          projection,
          this_week: {
            week_start: localYMD(thisWeek.weekStart),
            meters: thisWeek.meters,
            sessions: thisWeek.sessions,
          },
        });
        return;
      }
      console.log(
        `Progress: ${formatMeters(goal.totalMeters)} / ${formatMeters(goal.target)} (${((goal.progress ?? 0) * 100).toFixed(1)}%)`,
      );
      console.log(`Required pace: ${formatMeters(goal.requiredPace)} m/wk`);
      console.log(`Recent average: ${formatMeters(goal.currentAvgPace)} m/wk`);
      console.log(
        `Projection at current pace: ${formatMeters(projection.projected_total_meters)} m (${projection.projected_pct}%)`,
      );
      if (projection.shortfall_meters > 0) {
        console.log(`Projected shortfall: ${formatMeters(projection.shortfall_meters)} m`);
      } else {
        console.log("On track to exceed goal.");
      }
      console.log(
        `This week so far: ${formatMeters(thisWeek.meters)} m (${thisWeek.sessions} sessions)`,
      );
    });

  stats
    .command("splits <id>")
    .description("Split analysis for one workout (id or 'last')")
    .option("--json", "output as JSON")
    .action(async (ref: string, opts: { json?: boolean }) => {
      const cfg = await loadConfig();
      const workouts = await readWorkouts(dataPaths(cfg));
      const w = resolveWorkout(workouts, ref);
      if (w == null) {
        console.error(
          ref === "last" ? "No workouts found. Run `c2 sync` first." : `No workout with id ${ref}.`,
        );
        process.exit(1);
      }
      const rows = splitTable(w);
      const shape = splitShape(rows);

      if (opts.json) {
        printJSON("c2.stats.splits.v1", {
          workout_id: w.id,
          date: w.date,
          distance: w.distance,
          split_shape: shape,
          splits: rows,
        });
        return;
      }
      if (rows.length === 0) {
        console.log(`Workout ${w.id} has no split data.`);
        return;
      }
      console.log(`Workout ${w.id} — ${w.date} — ${formatMeters(w.distance)}m — ${shape} splits`);
      for (const s of rows) {
        console.log(
          `  ${s.index}: ${s.distance != null ? `${formatMeters(s.distance)}m` : "-"}  ${s.pace_500m ?? "-"}/500m  ${s.stroke_rate ?? "-"}spm  HR ${s.hr_avg ?? "-"}`,
        );
      }
    });

  stats
    .command("hr-pace")
    .description("Average heart rate by steady pace band (fitness proxy)")
    .option("-w, --weeks <n>", "window in weeks", "8")
    .option("--json", "output as JSON")
    .action(async (opts: { weeks: string; json?: boolean }) => {
      const weeks = parseWeeks(opts.weeks);
      const cfg = await loadConfig();
      const workouts = await readWorkouts(dataPaths(cfg));
      const bands = hrAtPace(workouts, new Date(), weeks);

      if (opts.json) {
        printJSON("c2.stats.hr-pace.v1", { weeks, bands });
        return;
      }
      if (bands.length === 0) {
        console.log("No steady workouts with heart rate data in the window.");
        return;
      }
      console.log(`Steady pace bands over the last ${weeks} weeks (HR avg, early→late half):`);
      for (const b of bands) {
        const trend =
          b.hr_delta == null
            ? ""
            : b.hr_delta < 0
              ? `  ↓${Math.abs(b.hr_delta)} (improving)`
              : b.hr_delta > 0
                ? `  ↑${b.hr_delta} (watch)`
                : "  → flat";
        const halves =
          b.early_avg_hr != null && b.late_avg_hr != null
            ? ` (${b.early_avg_hr}→${b.late_avg_hr})`
            : "";
        console.log(
          `  ${b.band}/500m: HR ${b.avg_hr}${halves} across ${b.workouts} workout${b.workouts === 1 ? "" : "s"}${trend}`,
        );
      }
    });
}

File: internal/cli/stats.go
package cli

import (
	"fmt"
	"strconv"
	"time"

	"github.com/spf13/cobra"

	"github.com/richhaase/c2/internal/analysis"
	"github.com/richhaase/c2/internal/display"
	"github.com/richhaase/c2/internal/envelope"
	"github.com/richhaase/c2/internal/models"
	"github.com/richhaase/c2/internal/stats"
	"github.com/richhaase/c2/internal/storage"
)

func jsNumber(v float64) string {
	return strconv.FormatFloat(v, 'f', -1, 64)
}

func orDash[T any](v *T, format func(T) string) string {
	if v == nil {
		return "-"
	}
	return format(*v)
}

type weeklyPayload struct {
	Weeks []stats.WeekSummaryData `json:"weeks"`
}

type goalPayload struct {
	Goal       stats.GoalProgress   `json:"goal"`
	Projection stats.GoalProjection `json:"projection"`
	ThisWeek   weekPayload          `json:"this_week"`
}

type splitsPayload struct {
	WorkoutID  int64               `json:"workout_id"`
	Date       string              `json:"date"`
	Distance   int                 `json:"distance"`
	SplitShape analysis.Shape      `json:"split_shape"`
	Splits     []analysis.SplitRow `json:"splits"`
}

type hrPacePayload struct {
	Weeks int                   `json:"weeks"`
	Bands []analysis.HRPaceBand `json:"bands"`
}

func newStatsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "stats",
		Short: "Derived training statistics",
	}
	cmd.AddCommand(
		newStatsWeeklyCmd(),
		newStatsGoalCmd(),
		newStatsSplitsCmd(),
		newStatsHRPaceCmd(),
	)
	return cmd
}

func newStatsWeeklyCmd() *cobra.Command {
	var (
		weeksFlag string
		asJSON    bool
	)
	cmd := &cobra.Command{
		Use:   "weekly",
		Short: "Weekly volume, sessions, pace, SPM, and HR",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			weeks, err := positiveInt(cmd, weeksFlag, "Error: --weeks must be a positive integer.")
			if err != nil {
				return err
			}
			_, p, err := loadStore()
			if err != nil {
				return err
			}
			workouts, err := storage.ReadWorkouts(p)
			if err != nil {
				return err
			}
			summaries := make([]stats.WeekSummaryData, 0, weeks)
			for _, ws := range stats.BuildWeekSummaries(workouts, time.Now(), weeks) {
				summaries = append(summaries, stats.WeekSummaryDataOf(ws))
			}

			out := cmd.OutOrStdout()
			if asJSON {
				return envelope.Print(out, "c2.stats.weekly.v1", weeklyPayload{Weeks: summaries})
			}
			fmt.Fprintln(out, "week        meters  sess  pace/500m   spm    hr")
			for _, s := range summaries {
				fmt.Fprintf(out, "%s  %8s  %4d  %9s  %4s  %4s\n",
					s.WeekStart,
					display.FormatMeters(s.Meters),
					s.Sessions,
					orDash(s.AvgPace500m, func(v string) string { return v }),
					orDash(s.AvgSPM, jsNumber),
					orDash(s.AvgHR, strconv.Itoa))
			}
			return nil
		},
	}
	cmd.Flags().StringVarP(&weeksFlag, "weeks", "w", "12", "number of weeks")
	cmd.Flags().BoolVar(&asJSON, "json", false, "output as JSON")
	return cmd
}

func newStatsGoalCmd() *cobra.Command {
	var asJSON bool
	cmd := &cobra.Command{
		Use:   "goal",
		Short: "Goal trajectory and projection",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			cfg, p, err := loadStore()
			if err != nil {
				return err
			}
			if cfg.Goal.StartDate == "" || cfg.Goal.EndDate == "" {
				return reportf(cmd, "Goal dates not configured. Run `c2 setup` to set start and end dates.")
			}
			workouts, err := storage.ReadWorkouts(p)
			if err != nil {
				return err
			}
			now := time.Now()
			goal, err := stats.ComputeGoalProgress(workouts, cfg, now)
			if err != nil {
				return err
			}
			end, err := configGoalEnd(cfg.Goal.EndDate)
			if err != nil {
				return err
			}
			projection := stats.ProjectGoal(goal, end, now)
			weeks := stats.RecentWeeks(workouts, now, 4)
			thisWeek := weeks[0]

			out := cmd.OutOrStdout()
			if asJSON {
				return envelope.Print(out, "c2.stats.goal.v1", goalPayload{
					Goal:       goal,
					Projection: projection,
					ThisWeek:   newWeekPayload(thisWeek.WeekStart, thisWeek.Meters, thisWeek.Sessions),
				})
			}

			fmt.Fprintf(out, "Progress: %s / %s (%s%%)\n",
				display.FormatMeters(goal.TotalMeters),
				display.FormatMeters(goal.Target),
				display.ToFixed(goal.Progress*100, 1))
			fmt.Fprintf(out, "Required pace: %s m/wk\n", display.FormatMeters(goal.RequiredPace))
			fmt.Fprintf(out, "Recent average: %s m/wk\n", display.FormatMeters(goal.CurrentAvgPace))
			fmt.Fprintf(out, "Projection at current pace: %s m (%s%%)\n",
				display.FormatMeters(projection.ProjectedTotalMeters), jsNumber(projection.ProjectedPct))
			if projection.ShortfallMeters > 0 {
				fmt.Fprintf(out, "Projected shortfall: %s m\n", display.FormatMeters(projection.ShortfallMeters))
			} else {
				fmt.Fprintln(out, "On track to exceed goal.")
			}
			fmt.Fprintf(out, "This week so far: %s m (%d sessions)\n",
				display.FormatMeters(thisWeek.Meters), thisWeek.Sessions)
			return nil
		},
	}
	cmd.Flags().BoolVar(&asJSON, "json", false, "output as JSON")
	return cmd
}

func newStatsSplitsCmd() *cobra.Command {
	var asJSON bool
	cmd := &cobra.Command{
		Use:   "splits <id>",
		Short: "Split analysis for one workout (id or 'last')",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			ref := args[0]
			_, p, err := loadStore()
			if err != nil {
				return err
			}
			workouts, err := storage.ReadWorkouts(p)
			if err != nil {
				return err
			}
			w := models.ResolveWorkout(workouts, ref)
			if w == nil {
				if ref == "last" {
					return reportf(cmd, "No workouts found. Run `c2 sync` first.")
				}
				return reportf(cmd, "No workout with id %s.", ref)
			}
			rows := analysis.SplitTable(*w)
			shape := analysis.SplitShape(rows)

			out := cmd.OutOrStdout()
			if asJSON {
				return envelope.Print(out, "c2.stats.splits.v1", splitsPayload{
					WorkoutID:  w.ID,
					Date:       w.Date,
					Distance:   w.Distance,
					SplitShape: shape,
					Splits:     rows,
				})
			}
			if len(rows) == 0 {
				fmt.Fprintf(out, "Workout %d has no split data.\n", w.ID)
				return nil
			}
			fmt.Fprintf(out, "Workout %d — %s — %sm — %s splits\n",
				w.ID, w.Date, display.FormatMeters(w.Distance), shape)
			for _, s := range rows {
				distance := "-"
				if s.Distance != nil {
					distance = display.FormatMeters(*s.Distance) + "m"
				}
				fmt.Fprintf(out, "  %d: %s  %s/500m  %sspm  HR %s\n",
					s.Index,
					distance,
					orDash(s.Pace500m, func(v string) string { return v }),
					orDash(s.StrokeRate, strconv.Itoa),
					orDash(s.HRAvg, strconv.Itoa))
			}
			return nil
		},
	}
	cmd.Flags().BoolVar(&asJSON, "json", false, "output as JSON")
	return cmd
}

func newStatsHRPaceCmd() *cobra.Command {
	var (
		weeksFlag string
		asJSON    bool
	)
	cmd := &cobra.Command{
		Use:   "hr-pace",
		Short: "Average heart rate by steady pace band (fitness proxy)",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			weeks, err := positiveInt(cmd, weeksFlag, "Error: --weeks must be a positive integer.")
			if err != nil {
				return err
			}
			_, p, err := loadStore()
			if err != nil {
				return err
			}
			workouts, err := storage.ReadWorkouts(p)
			if err != nil {
				return err
			}
			bands := analysis.HRAtPace(workouts, time.Now(), weeks)

			out := cmd.OutOrStdout()
			if asJSON {
				if bands == nil {
					bands = []analysis.HRPaceBand{}
				}
				return envelope.Print(out, "c2.stats.hr-pace.v1", hrPacePayload{Weeks: weeks, Bands: bands})
			}
			if len(bands) == 0 {
				fmt.Fprintln(out, "No steady workouts with heart rate data in the window.")
				return nil
			}
			fmt.Fprintf(out, "Steady pace bands over the last %d weeks (HR avg, early→late half):\n", weeks)
			for _, b := range bands {
				trend := ""
				if b.HRDelta != nil {
					switch {
					case *b.HRDelta < 0:
						trend = fmt.Sprintf("  ↓%d (improving)", -*b.HRDelta)
					case *b.HRDelta > 0:
						trend = fmt.Sprintf("  ↑%d (watch)", *b.HRDelta)
					default:
						trend = "  → flat"
					}
				}
				halves := ""
				if b.EarlyAvgHR != nil && b.LateAvgHR != nil {
					halves = fmt.Sprintf(" (%d→%d)", *b.EarlyAvgHR, *b.LateAvgHR)
				}
				plural := "s"
				if b.Workouts == 1 {
					plural = ""
				}
				fmt.Fprintf(out, "  %s/500m: HR %d%s across %d workout%s%s\n",
					b.Band, b.AvgHR, halves, b.Workouts, plural, trend)
			}
			return nil
		},
	}
	cmd.Flags().StringVarP(&weeksFlag, "weeks", "w", "8", "window in weeks")
	cmd.Flags().BoolVar(&asJSON, "json", false, "output as JSON")
	return cmd
}
